Add Kafka transactions blog post (#362)
strimzi/strimzi.github.io: _posts/2023-04-14-kafka-transactions.md

---
layout: post
title: "Kafka transactions (exactly-once semantics)"
date: 2023-04-14
author: federico_valeri
---

> 

Reviewer: You've got some good context here, but what's missing in the first two or three paragraphs is telling the reader what they're going to learn from the rest of this post. What is the point you're trying to inform them about/why should they care?

Author: I'll try to improve that. In addition to learn how it works, I thinks drawbacks and how the txn metadata is stored, which helps in troubleshooting.

Thanks.

Apache Kafka is a distributed streaming platform that allows efficient and reliable processing of messages.
It is widely used in various industries for data streaming applications, such as processing real-time data, event sourcing, and microservices integration.
One of the key features of Kafka is the support for transactions, which provides exactly-once semantics (EOS) and is available since Kafka 0.11 ([KIP-98](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging)).
The support for EOS was recently extended to source connectors in Kafka Connect, which enables configuration-driven and fully transactional streaming pipelines ([KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors)).

As messages flow through Kafka, they can be processed by multiple consumers, and ensuring that all consumers that should receive a message actually receive it once and only once can be crucial.

> The EOS guarantees that any set of operations inside Kafka will be completed atomically, either all of the operations succeed or none of them do.

Author: I agree, but this is only the intro and I'm clarifying this point in the considerations. There, I'll replace "external systems" with "processors with externally observable side-effects", because simply accessing an external system does not mean changing its state. Thanks.

This is done by defining transaction boundaries that dictate which operations can be grouped together in a single transaction.
These operations can include reading from topics, writing to topics, and committing offsets.

> A failure during processing can cause a transaction to be aborted, in which case none of the messages from the transaction will be readable by consumers.

Reviewer 2: Only for consumers that set their isolation to `read_committed`

This helps in maintaining high levels of reliability and consistency as data is processed throughout the application's lifetime.

<!--more-->

### Delivery guarantees

In distributed systems like Kafka, any component can fail independently and you have to deal with this fact.
A broker may crash or a network failure may happen while the producer is sending a message to a topic.

Depending on the configuration, you can have three different delivery semantics:

- **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)

> - **at-least-once**: the producer retries but the broker can fail after writing but before sending the ack (duplicates risk)

Author: It's scary like this, isn't it? Fixed. Thanks

- **exactly-once**: even if the producer retries sending a message, it is written exactly once (idempotency)

With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. financial institutions).

> To be fair, there is no such thing as exactly-once delivery in distributed systems (two generals problem), but we can fake it by implementing idempotent operations.

Author: Sure.

Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.

> Since Kafka 3.0, the producer enables the strongest delivery guarantees by default (`acks=all`, `enable.idempotence=true`).

Reviewer 2: I think the transactional producer gives the strongest semantics. With idempotency these guarantees are only valid per message and for the lifetime of the producer instance.

In addition to durability, this provides partition level message ordering and duplicates protection.

When the idempotent producer is enabled, the broker registers the producer id (PID) and the epoch (producer session).
A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
The broker keeps track of the sequence number per producer and partition, periodically storing this information in `.snapshot` files.
Whenever a new batch arrives, the broker checks if the last received number is equal to the first batch number plus one, then the batch is acknowledged, otherwise it is rejected.

Unfortunately, the idempotent producer does not guarantee atomicity when you need to write to multiple partitions as a single unit of work.
This is the typical scenario with many stream processing applications that run read-process-write cycles.
In this case, you can use transactions API or Streams API to make these cycles atomic.
The transaction API is low level and you need to ensure the right configuration and to commit offsets within the scope of the transaction.

> Instead, with the Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.

Author: Agree.

### Considerations

> Each producer must configure its own static and unique `transactional.id` (TID) to enable the zombie fencing logic, which avoids message duplicates.

Reviewer 2: What about linking `transactional.id` to the official docs?
https://kafka.apache.org/documentation/#producerconfigs_transactional.id

Author: Agree to both. This is a critical configuration and we don't want ambiguity.

The TID is used to uniquely identify the same producer instance across process restarts, and it is associated with the producer session.

> When starting, the producer must call `initTransactions` one time, so that the broker can complete any open transaction with the given TID and increment the epoch.

Author: Ok.

> Once the epoch is incremented, any producers with the same TID and an older epoch are considered zombies, and are fenced off as soon as they try to send data.

Author: Better.

> Before Kafka 2.6 the TID had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.

> > This limitation was ugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 

Author: Fixed.

> > This limitation was ugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 

Author: Fixed.

To guarantee message ordering, a given producer can have at most one open transaction.
Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.

The `TransactionCoordinator` is a module running inside every Kafka broker.
Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.

It is important to note that transactions are only supported inside a single Kafka cluster.
If the transaction scope includes external systems, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.
This component is able to synchronize a database transaction with the Kafka transaction.

#### Transaction states

A transaction goes through the following states:

1. **Undecided** (Ongoing)
2. **Decided and unreplicated** (PrepareCommit\|PrepareAbort)
3. **Decided and replicated** (CompleteCommit\|CompleteAbort)

> Non-transactional batches are considered decided immediately, but transactional batches are only decided when the corresponding commit or abort marker is written (control record).

Author: Right, I need to add more context here.

The high watermark (HW) is the offset of the last message that was successfully committed (replicated to all in-sync replicas).
The first unstable offset (FUO) is the earliest offset that is part of the ongoing transaction, if present.
The last stable offset (LSO) is the offset such that all lower offsets have been decided and it is always present.

> The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW (LSO <= HW <= LEO).

Author: It's in the image, but I plan to rephrase and add more details.

> The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW (LSO <= HW <= LEO).

Author: Yes, I'll add one more line, putting that in a code block.

![txn log](/assets/images/posts/2023-04-14-kafka-txn-log.png)

#### Error handling

If Kafka authorization is enabled, you need to configure the appropriate ACLs rules to avoid `TransactionalIdAuthorizationException`.

In the following example we allow any `transactional.id` from any user.

```sh
$ kafka-acls.sh --bootstrap-server:9092 --command-config client.properties --add \
  --allow-principal User:* --operation write --transactional-id *
```

If any of the send operations inside the transaction scope fails, the final `commitTransaction` will fail and throw the exception from the last failed send.
When this happens, your application should call `abortTransaction` to reset the state and apply a retry strategy.

> Some errors cannot be recovered and the application should terminate (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`).

Reviewer 2: Or close the current KafkaProducer instance and create a new one?

Author: Yeah, termination is too drastic if you are not running on K8s or need fast recovery times.

> Some errors cannot be recovered and the application should terminate (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`).

Author: Isn't the error handling of the basic example presented below good enough? In that case, I could refer to it. In order to keep it simple, there is no smart retry strategy (e.g. fixed retries or exponential backoff), but it should be correct.

#### Monitoring

When running in production, it is important to monitor the health of your Kafka deployments to maintain reliable performance.
Some of the metrics that Kafka provides can be useful when dealing with transaction issues and tuning.

> On the broker side, you should should collect at least the count and 99 percentile for the following APIs.

Author: I agree thanks.

```sh
# requests: FindCoordinator, InitProducerId, AddPartitionsToTxn, Produce, EndTxn
kafka.network:type=RequestMetrics,name=RequestsPerSec,request=([\w]+)
kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
attributes:Count,99thPercentile
```

> Producer metrics are more difficult to collect, but they help when broker logs reveal some misbehavior.

Author: Thanks.

```sh
kafka.producer:type=producer-metrics,client-id=([-.\w]+)
attributes:txn-init-time-ns-total,txn-begin-time-ns-total,txn-send-offsets-time-ns-total,txn-commit-time-ns-total,txn-abort-time-ns-total
```

> ### Drawbacks

Author: I think we can skip that as the whole post is dedicated to transactions.

There are a few drawbacks when using Kafka transactions that developers should be aware of.

> Firstly, transactions can add some latency to an application due to the write amplification required to ensure EOS.

Author: That's a very good point that I forgot to include. Thanks.

While this may not be a significant issue for some applications, those that require real-time processing may be affected.

For each transaction we have the following overhead, which is independent of the number of messages: 

- some RPCs to register the partitions with the coordinator (can be batched)
- one marker has to be written to each partition when completing a transaction

> - few state change records are appended to the internal transaction state log

Author: Much better. Thanks.

Most of this latency is on the producer side, where the transactions API is implemented.
The performance degradation can be significant when having short transaction commit intervals.

> Increasing the transaction duration also increases the end-to-end latency, so it's a matter of tradeoff.

Reviewer 2: Plus the risk an issue is hit and the whole transaction has to be aborted.

Author: Yes, it may also require additional tuning.

> Another potential drawback is hanging transactions, which are transactions with a missing or out of order control record.

Author: Yeah, we don't want to scare the reader, but it's important to be aware and set alerts, because it can impact service availability.

They may be caused by a network issue causing delayed messages (see [KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).

A hanging transaction is usually revealed by a stuck application with growing lag.
Despite messages being produced, consumers can't make any progress because the LSO does not grow anymore.

> ```sh

Reviewer 2: Maybe add a line like:
```
For example, you will see the following messages in the consumer logs:
```

[Consumer clientId=my-client, groupId=my-group] The following partitions still have unstable offsets which are not cleared on the broker side: [__consumer_offsets-27], 
this could be either transactional offsets waiting for completion, or normal offsets waiting for replication after appending to local log
```

Additionally, if the partition belongs to a compacted topic, this causes the unbounded partition growth, because the `LogCleaner` does not clean beyond the LSO.
If not detected and fixed in time, this may lead to disk space exhaustion and service disruption.

> Fortunately, there are embedded tools that can be used to find all hanging transactions and rollback them.

Reviewer 3: ```suggestion
Fortunately, there are embedded tools that can be used to find all hanging transactions and roll them back.
```

> Fortunately, there are embedded tools that can be used to find all hanging transactions and rollback them.

Author: Right. There are useful information there.

After that, the LSO starts to increment again on every transaction completion and the application should be able to recover.

```sh
# find all hanging transactions in a broker

> $ kafka-transactions.sh --bootstrap-server :9092 find-hanging --broker 0

Reviewer 2: Maybe put `./bin/` in front

Author: I think I will use `$KAFKA_HOME/bin/` everywhere, as it doesn't require to mess with the reader's PATH.

Topic                  Partition   ProducerId  ProducerEpoch   StartOffset LastTimestamp               Duration(s)
__consumer_offsets     27          171100      1               913095344   2022-06-06T03:16:47Z        209793

# abort an hanging transaction

> $ kafka-transactions.sh --bootstrap-server :9092 abort --topic __consumer_offsets --partition 27 --start-offset 913095344

Author: Yes, better to be explicit about this point. Thanks.

```

### Basic example

> Let's now run a basic example to see how transactions work at the partition level.

Reviewer 3: ```suggestion
Let's run a basic example to see how transactions work at the partition level.
```

Author: Ok.

> You just need a computer with few essential command line tools.

Reviewer 2: `computer` -> `environment`
Maybe also list the tools we're going to use

> This application consumes text messages from the `input-topic`, reverts their content and sends the result to the `output-topic`.

Reviewer 3: ```suggestion
This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.
```

> ![txn app](/assets/images/posts/2023-04-14-kafka-txn-app.png)

Author: I agree that RPCs are better here. Will update the diagram.

> In the following snippet we start the cluster, run the application and create a single transaction.

Reviewer 3: ```suggestion
In the following snippet we start the cluster, run the application, and create a single transaction.
```

Author: Ok.

```sh
# download the Apache Kafka binary and start a single-node Kafka cluster in background
$ KAFKA_VERSION="3.4.0" KAFKA_HOME="$(mktemp -d -t kafka.XXXXXXX)" PATH="$KAFKA_HOME/bin:$PATH" && export KAFKA_HOME PATH

$ curl -sLk "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.13-$KAFKA_VERSION.tgz" \
  | tar xz -C "$KAFKA_HOME" --strip-components 1

$ zookeeper-server-start.sh -daemon $KAFKA_HOME/config/zookeeper.properties \
  && sleep 5 && kafka-server-start.sh -daemon $KAFKA_HOME/config/server.properties

# open a new terminal, get and run the application (Ctrl+C to stop)
$ git clone git@github.com:fvaleri/examples.git -n --depth=1 --filter=tree:0 && cd examples \
  && git sparse-checkout set --no-cone kafka/kafka-txn && git checkout 5907e0bd31175b5f411dc0e0804cf69cde76ec90
...
HEAD is now at 5907e0b Add SerializationException

$ export BOOTSTRAP_SERVERS="localhost:9092" INSTANCE_ID="kafka-txn-0" \
  GROUP_ID="my-group" INPUT_TOPIC="input-topic" OUTPUT_TOPIC="output-topic"

$ mvn clean compile exec:java -q -f kafka/kafka-txn/pom.xml
Starting instance with TID kafka-txn-0
Created topics: input-topic
Waiting for new data
...

> # come back to the previus terminal and send a test message

Reviewer 2: `previus` -> `previous`

$ kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
>test
>^C
```

> All of this is pretty boring, but now comes the interesting part.

Reviewer 2: `boring` -> `normal`/`classic` or even only keep `Now comes the interesting part.`

How can we identify which partitions are involved in this transaction and inspect their content?

> We are writing to the `output-topic` which has only one partition, but the internal topics for storing CG offsets and transaction states have 50 partitions each by default.

Reviewer 2: `CG` -> `consumer group`

> To avoid looking at 100 partitions in the worst case scenario, we can use the same hashing function that Kafka uses to find the coordinating partition.

Author: Here we are talking about two internal topics and we want to look at both, so 100 partitions in the worst case!

> To avoid looking at 100 partitions in the worst case scenario, we can use the same hashing function that Kafka uses to find the coordinating partition.

Author: Ok.

| offset: 1 CreateTime: 1680383688085 keySize: 26 valueSize: 24 sequence: 0 headerKeys: [] key: offset_commit::group=my-group,partition=input-topic-0 payload: offset=1
baseOffset: 2 lastOffset: 2 count: 1 baseSequence: -1 lastSequence: -1 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: true deleteHorizonMs: OptionalLong.empty position: 457 CreateTime: 1665506597998 size: 78 magic: 2 compresscodec: none crc: 3355926470 isvalid: true
| offset: 2 CreateTime: 1680383688163 keySize: 4 valueSize: 6 sequence: -1 headerKeys: [] endTxnMarker: COMMIT coordinatorEpoch: 0
...
```

As you can see, the consumer offsets and the output message are committed as an atomic unit.
In `wc-output-0`, we see that the data batch is transactional and contains the producer's state (PID and epoch).
This data batch is followed by a control batch, which contains the end transaction marker (COMMIT).

> In `__consumer_offsets-12`, the consumer group offset commit batch is followed by a similar control batch.

Reviewer: The output is really dense, so I don't think readers will necessarily follow along at this point. You could reformat the output so you can here easily refer to specific lines to highlight to the reader what they're looking for.

Additionally, using Kafka transactions can help ensure that the ordering of events is preserved, which can be necessary for certain applications.

In this post you will learn how EOS works in Kafka, which are the main components that are involved in a transaction lifetime and their requirements. 
Then, you will learn about the drawbacks, so that you can anticipate any issue long before your application reach production stage.
Finally, you will see a basic example that helps to understand how transactions metadata are stored, which is useful for troubleshooting.

A general understanding of how the log replication protocol works is required to fully understand some of the concepts presented here.

<!--more-->

### Delivery guarantees

In distributed systems like Kafka, any component can fail independently and you have to deal with this fact.
For example, a broker may crash or a network failure may happen while the producer is sending a message to a topic.

Depending on the configuration, you can have three different delivery semantics:

> - **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)

Author: Well, you could have at-most-once semantics also when waiting for the ack. This depends on both the config and app logic, so I will clarify. Thanks.

> - **at-most-once**: the producer does not retry when there is a timeout or error (data loss risk)

Author: Clarified in a later comment.

With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. a duplicate debit from a bank account).

To be fair, there is no such thing as exactly-once delivery in distributed systems ([two generals' problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)), but we can fake it by implementing idempotent log append.
Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.

> Idempotence is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.

Since Kafka 3.0, the producer enables the stronger delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
In addition to durability, this provides partition level message ordering and duplicates protection.

> When the idempotent producer is enabled, the broker registers the producer id (PID) and the epoch (producer session).

Author: I agree, this is not relevant and misleading. Thanks.

A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
The broker keeps track of the sequence number per producer and partition, periodically storing this information in a `.snapshot` file.
Whenever a new batch arrives, the broker checks if the last received number is equal to the first batch number plus one, then the batch is acknowledged, otherwise it is rejected.

Unfortunately, the idempotent producer does not guarantee duplicate protection across restarts and when you need to write to multiple partitions as a single unit of work.
This is the typical scenario with many stream processing applications that run read-process-write cycles.
In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.

> The Kafka Producer API is low level, so you need to ensure the right configuration and to commit offsets within the scope of the transaction.

Reviewer: This is correct, but I'm not sure how many readers would really understand what it means in practice. If you don't then it's a bit frustrating because you don't elaborate on what the right configuration is, or what commit within transaction scope means. So I wonder if we can use some more general language rather than this paragraph. How about:

> The transaction support exposed by the Producer API is low level, and needs to be used with carefully in order to actually get transactional semantics. For those using Kafka Streams it is much simpler: setting `processing.guarantee=exactly_once_v2` will enable EOS on your existing topology.

Author: Agree and the basic example shows the correct way of using the Producer API.

A transaction should be aborted in case failure during processing, which means none of the messages from the transaction will be readable by transactional consumers.
Instead, with the Kafka Streams API you can simply set `processing.guarantee=exactly_once_v2` to enable EOS on your existing topology.

### Considerations

Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.

> The `transactional.id` is used to uniquely identify the same producer instance across process restarts, and it is associated with the producer session.

Reviewer: ```suggestion
The `transactional.id` is used to uniquely identify the same logical producer across process restarts. In contrast to the idempotent producer, a transactional producer instance will be allocated _the same_ PID (but incremented producer epoch) as any previous instance with the same `transactional.id`. Together the PID and producer epoch identify a logical _producer session_.
```

Author: Thanks! This really makes the producer session concept clear.

> When starting, the producer must call `initTransactions` one time, so the broker can tidy up the state associated with any previous incarnations of this producer, as identified by the `transactional.id`.

Reviewer: ```suggestion
When starting, the producer call `initTransactions` to initialize t producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
```

Author: Ok.

Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.

> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 
> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).

To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).
Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.

> The transaction coordinator is a module running inside every Kafka broker.

Reviewer: Strictly speaking it's only brokers which are leading a partition of `__transaction_state`. In small clusters this ends up being every broker. Rather than contort the language to fit the facts (and risk getting side tracked), let's just be a bit vague.

```suggestion
The transaction coordinator is a module running inside a Kafka broker.
```

Author: Ok.

Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).
The coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.

It is important to note that transactions are only supported inside a single Kafka cluster.

> If the transaction scope includes processors with externally observable side-effects, you would need to use an additional component such as `ChainedKafkaTransactionManager` from the SpringBoot project.

Author: I'll drop this reference.

The ChainedKafkaTransactionManager is deprecated. Spring still supports this, but I don't have much experience with that and it would require a dedicated post.

Instead, I'll mention the "transactional outbox pattern" (with a link), which addresses the same concern and it's implementation independent.

> This component is able to synchronize a database transaction with the Kafka transaction.

Reviewer: Well I think that comes with some caveats, so maybe we should reword this claim to reflect what it actually does.

Author: I'll drop this reference.

The ChainedKafkaTransactionManager is deprecated. Spring still supports this, but I don't have much experience with that and it would require a dedicated post.

Instead, I'll mention the "transactional outbox pattern" (with a link), which addresses the same concern and it's implementation independent.

#### Transaction metadata

> A control record is a special type of record that is used for internal operations within Kafka, rather than for normal message processing.

Reviewer: You need a sentence here to explain that you're now talking about the structure of the log that a broker keeps on disk. 

> In order to implement transactions Kafka brokers need to include extra book-keeping information in logs. Information about producers and their transactions is stored in the `__transaction_state` log which is appended to by the transaction coordinator.
Within user logs in addition to the usual batches of user records, partition leaders append control records to track which batches have actually been committed and which rolled back.
Control records are exposed to Kafka applications by Kafka consumers, as they are an internal implementation detail of the transaction support.

Author: Thanks.

Control records are typically not visible to Kafka consumers, as they are meant to be processed only by the Kafka brokers.

An end transaction marker (`COMMIT` or `ABORT`) is a special control record that indicates the end of a transaction (decision).
It is used to ensure that the transaction is either completed successfully or aborted in case of a failure.
The end transaction marker is added to the log after all the records related to the transaction have been written to the log.

Non-transactional records are considered decided immediately, but transactional records are only decided when the corresponding commit or abort marker is written.

A transaction goes through the following states:

1. **Undecided** (Ongoing)
2. **Decided and unreplicated** (PrepareCommit\|PrepareAbort)
3. **Decided and replicated** (CompleteCommit\|CompleteAbort)

This is handled through a process similar to the two-phase commit protocol.
Every time some data is sent to a new partition inside a transaction, we have a new `Ongoing` control record.
When the transaction completes, we have the `PrepareCommit` or `PrepareAbort` control record and the transaction coordinator is doomed to commit or abort.

> The final step consists of writing the end transaction markers to each involved partition, and the `CompleteCommit` or `CompleteAbort` control record.

Reviewer: This could be improved by explaining that these states are from the transaction coordinator's PoV. Right now you're blurring commit markers in user partitions with the (persistent) state in the coordinator, and that will be confusing to anyone who doesn't know this stuff already.

Author: You are right. That could be confusing, so let me add a line here.

![txn log](/assets/images/posts/2023-04-14-kafka-txn-log.png)

> The last stable offset (LSO) is the offset such that all lower offsets have been decided and it is always present.

Reviewer: ```suggestion
The last stable offset (LSO) is the offset in a user partition such that all lower offsets have been decided and it is always present.
```

Author: Ok.

> The first unstable offset (FUO) is the earliest offset that is part of the ongoing transaction, if present.

Reviewer: ```suggestion
The first unstable offset (FUO) is the earliest offset in a user partition that is part of the ongoing transaction, if present.
```

Author: Ok.

> The LSO is equal to the FUO if it's lower than the HW, otherwise it's the HW.

Reviewer: equal to, or `LSO = FUO -1`?

Reviewer: My point was the +1 is missing. Currently you're saying that "The LSO is equal to the FUO", i.e. the offset is both Stable and also Unstable.

We have that `LSO <= HW <= LEO`, which can be the same offset when there is no open transaction and all partitions are in-sync.

> #### Error handling

Reviewer: I'd change this to `#### Authorization`

Author: Ok.

The `TransactionalIdAuthorizationException` happens when you have authorization enabled without the appropriate ACLs rules.

In the following example we allow any `transactional.id` from any user.

```sh
$ $KAFKA_HOME/bin/kafka-acls.sh --bootstrap-server :9092 --command-config client.properties --add \
  --allow-principal User:* --operation write --transactional-id *
```

> If any of the send operations inside the transaction scope fails, the final `commitTransaction` will fail and throw the exception from the last failed send.

Reviewer: ...And insert a new heading `#### Error handling` here

Author: Ok.

When this happens, your application should call `abortTransaction` to reset the state and apply a retry strategy.

> Some errors cannot be recovered (`ProducerFencedException`, `FencedInstanceIdException`, `OutOfOrderSequenceException`) and the application should close the current producer instance and create a new one.

Reviewer: How does the reader know which?

Author: I think you need to set a producer callback and log in case of error.

This is where frameworks can provide useful abstractions over multiple producer instances.

Then, transactions can add some latency to an application due to the write amplification required to ensure EOS.
While this may not be a significant issue for some applications, those that require real-time processing may be affected.

Each transaction in Kafka incurs some overhead that is independent of the number of messages involved. 
This overhead includes the following:

- A small number of Remote Procedure Calls (RPCs) to register the partitions with the coordinator. 
  These RPCs can be batched together to reduce overhead.
- A marker that must be written to each partition when the transaction is completed.
- A few state change records that are appended to the internal transaction state log.

Most of this latency is on the producer side, where the transactions API is implemented.
The performance degradation can be significant when having short transaction commit intervals.
Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.

> 

Reviewer: It's worth mentioning the impacts on the `isolation.level=read_committed` consumer too. Although it will only be able to fetch upto the LSO it still has to buffer records in memory until is observes the commit marker, so you expect to see increased memory usage.

Author: That's true. Thanks for pointing this out.

Another potential drawback is hanging transactions, which are transactions with a missing or out-of-order control record.

> They may be caused by a network issue causing delayed messages ([KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).

Reviewer: It's probably worth having a separate section for hanging transaction. It's not a tradeoff like the other drawbacks mentioned.

This issue is rare, but it's important to be aware and set alerts, because it can impact service availability.

A hanging transaction is usually revealed by a stuck application with growing lag ([KIP-664](https://cwiki.apache.org/confluence/display/KAFKA/KIP-664%3A+Provide+tooling+to+detect+and+abort+hanging+transactions)).

> Despite messages being produced, consumers can't make any progress because the LSO does not grow anymore.

Reviewer: You should mention here about transaction timeout, and how hanging transactions are different.

Author: I would say:

The transaction coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.
This mechanism does not work for hanging transactions because they are marked as completed.

```

Additionally, if the partition belongs to a compacted topic, this causes the unbounded partition growth, because the `LogCleaner` does not clean beyond the LSO.
If not detected and fixed in time, this may lead to disk space exhaustion and service disruption.

Fortunately, there are embedded tools that can be used to find all hanging transactions and roll them back.
When doing that, it is important to understand the business context surrounding the records in this transaction and possibly append those records again using a new transaction.

```sh
# find all hanging transactions in a broker
$ $KAFKA_HOME/bin/kafka-transactions.sh --bootstrap-server :9092 find-hanging --broker 0
Topic                  Partition   ProducerId  ProducerEpoch   StartOffset LastTimestamp               Duration(s)
__consumer_offsets     27          171100      1               913095344   2022-06-06T03:16:47Z        209793

# abort an hanging transaction
$ $KAFKA_HOME/bin/kafka-transactions.sh --bootstrap-server :9092 abort --topic __consumer_offsets --partition 27 --start-offset 913095344
```

> ### Basic example

Author: I would prefer to keep it last. I don't think you can really predict how a reader will consume this content, because it mostly depends on prior knowledge. I do expect some back and forth between theory and example. The impatient reader may want to jump straight into the example and then come back to the theory.

> ### Basic example

Reviewer 4: I am not sure to agree ... or maybe this article is for people who already have a good understand of transactions already.

$ $KAFKA_HOME/bin/kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
>test
>^C
```

All of this is pretty normal, but now comes the interesting part.

How can we identify which partitions are involved in this transaction and inspect their content?
We are writing to the `output-topic` which has only one partition, but the internal topics for storing consumer group offsets and transaction states have 50 partitions each by default.

To avoid looking at 100 partitions in the worst case scenario, we can use the same hashing function that Kafka uses to find the coordinating partition.
This function maps `group.id` to `__consumer_offsets` partition, and `transactional.id` to `__transaction_state` partition.

```sh
# define the function and find the coordinating partitions
$ kafka-cp() {
  local id="${1-}" part="${2-50}"
  if [[ -z $id ]]; then echo "Missing id parameter" && return; fi
  echo 'public void run(String id, int part) { System.out.println(abs(id.hashCode()) % part); }
    private int abs(int n) { return (n == Integer.MIN_VALUE) ? 0 : Math.abs(n); }
    run("'"$id"'", '"$part"');' | jshell -
}

$ kafka-cp my-group

> 12

Reviewer: Can we annotate this is a comment `# to the applications offsets are in partition 12 of __consumer_offsets`?

$ kafka-cp kafka-txn-0

> 30

Reviewer: Can we annotate this is a comment `# to the application's transactional state is in partition 30 of __transaction_state`?

Author: Ok.

12 | offset: 3 CreateTime: 1680383688149 keySize: 17 valueSize: 87 sequence: -1 headerKeys: [] key: transaction_metadata::transactionalId=kafka-txn-0 payload: producerId:0,producerEpoch:0,state=PrepareCommit,partitions=[output-topic-0,__consumer_offsets-12],txnLastUpdateTimestamp=1680383688148,txnTimeoutMs=60000
13 baseOffset: 4 lastOffset: 4 count: 1 baseSequence: -1 lastSequence: -1 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false isControl: false deleteHorizonMs: OptionalLong.empty position: 615 CreateTime: 1680383688180 size: 122 magic: 2 compresscodec: none crc: 3020616838 isvalid: true
14 | offset: 4 CreateTime: 1680383688180 keySize: 17 valueSize: 37 sequence: -1 headerKeys: [] key: transaction_metadata::transactionalId=kafka-txn-0 payload: producerId:0,producerEpoch:0,state=CompleteCommit,partitions=[],txnLastUpdateTimestamp=1680383688154,txnTimeoutMs=60000
```

> In `__transaction_state-20` dump, we can see all state changes keyed by `transactionalId` (lines 6, 8, 10, 12, 14).

Author: Thanks for spotting this. I updated the instance id and forgot to fix some references.

---
layout: post

> title: "Kafka transactions (exactly-once semantics)"

Author: I like it. Thanks.

date: 2023-04-14
author: federico_valeri
---

Apache Kafka is a distributed streaming platform that allows efficient and reliable processing of messages.
It is widely used in various industries for data streaming applications, such as processing real-time data, event sourcing, and microservices integration.
One of the key features of Kafka is the support for transactions, which provides exactly-once semantics (EOS) and is available since Kafka 0.11 ([KIP-98](https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging)).

> The support for EOS was recently extended to source connectors in Kafka Connect, which enables configuration-driven and fully transactional streaming pipelines ([KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors)).

Author: No direct relation, I just wanted to highlight the fact that you can have a transactional pipeline without having to write a single line of code.

> The support for EOS was recently extended to source connectors in Kafka Connect, which enables configuration-driven and fully transactional streaming pipelines ([KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors)).

Reviewer 4: Are we sure that "configuration-driven" is clear? It wasn't for me at least. You explained much better with these words now :-)

Transactions provide the guarantee of atomicity, consistency, isolation, and durability (ACID) for a group of messages that are published to Kafka.
This means that either all messages within the transaction will be successfully written to Kafka or none of the messages will be written.
This is a crucial feature when it comes to ensuring data consistency and avoiding any data loss.

> Kafka transactions are particularly important in scenarios where data integrity is critical, such as financial transactions, logging systems, and messaging platforms.

Author: Agree. Removed.

Additionally, using Kafka transactions can help ensure that the ordering of events is preserved, which can be necessary for certain applications.

In this post you will learn how EOS works in Kafka, which are the main components that are involved in a transaction lifetime and their requirements. 
Then, you will learn about the drawbacks, so that you can anticipate any issue long before your application reach production stage.
Finally, you will see a basic example that helps to understand how transactions metadata are stored, which is useful for troubleshooting.

> A general understanding of how the log replication protocol works is required to fully understand some of the concepts presented here.

Author: Ok. Done.

With at-least-once, there is no way for the producer to know the nature of the failure, so it simply assumes that the message was not written and retries.
There are use cases in which lost or duplicated messages are acceptable, but in others this would cause serious business problems (e.g. a duplicate debit from a bank account).

To be fair, there is no such thing as exactly-once delivery in distributed systems ([two generals' problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem)), but we can fake it by implementing idempotent log append.
Another way to achieve EOS is to apply a deduplication logic at the consumer side, but this is a stateful logic that needs to be replicated in every consumer.

> Idempotence is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.

Since Kafka 3.0, the producer enables the stronger delivery guarantees by default (`acks=all`, `enable.idempotence=true`).
In addition to durability, this provides partition level message ordering and duplicates protection.

> When the idempotent producer is enabled, the broker registers a producer id (PID) for every new producer instance, including restarts.

Author: It means that an application restart would create a new producer instance. It's redundant and confusing. Let me remove it.

A sequence number is assigned to each record when the batch is first added to a produce request and never changed, even if the batch is resent.
The broker keeps track of the sequence number per producer and partition, periodically storing this information in a `.snapshot` file.
Whenever a new batch arrives, the broker checks if the received sequence number is equal to the last-appended batch's sequence number plus one, then the batch is acknowledged, otherwise it is rejected.

Unfortunately, the idempotent producer does not guarantee duplicate protection across restarts and when you need to write to multiple partitions as a single unit of work.
This is the typical scenario with many stream processing applications that run read-process-write cycles.
In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.

The transaction support exposed by the Producer API is low level, and needs to be used carefully in order to actually get transactional semantics.
For those using Kafka Streams it is much simpler: setting `processing.guarantee=exactly_once_v2` will enable EOS on your existing topology.

### Considerations

> Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.

Author: There is no need to mention zombie fencing in the opening line, as it is explained few lines below.

In contrast to the idempotent producer, a transactional producer instance will be allocated the same PID (but incremented producer epoch), as any previous instance with the same `transactional.id`. 
Together the PID and producer epoch identify a logical producer session.

> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition.
> It was fixed by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).

When starting, the client application calls `Producer.initTransactions` to initialize the producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.

> Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.

Author: There is no need to mention zombie fencing in the opening line, as it is explained few lines below.

To guarantee message ordering, a given producer can have at most one ongoing transaction (they are executed serially).
Brokers will prevent consumers with `isolation.level=read_committed` from advancing to offsets which include open transactions.
Messages from aborted transactions are filtered out within the consumer, and are never observed by the client application.

It is important to note that transactions are only supported inside a single Kafka cluster.
If the transaction scope includes processors with externally observable side-effects, you would need to use additional components.
The [transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) could help in this case.

#### Transaction metadata

In order to implement transactions Kafka brokers need to include extra book-keeping information in logs.

Information about producers and their transactions is stored in the `__transaction_state` topic which is used by the transaction coordinator running inside a broker.

> Each transaction coordinator owns a subset of the partitions in the `__transaction_state` topic (i.e. the partitions for which its broker is the leader).

Author: It's in the line above this, which I reworded a bit.

Within user logs in addition to the usual batches of data records, transaction coordinators cause partition leaders append control records (commit/abort markers) to track which batches have actually been committed and which rolled back.

> Control records are not exposed to applications by Kafka consumers, as they are an internal implementation detail of the transaction support.

Author: Well, that would not be correct, because the Kafka consumer buffers them in order to know which transactions are aborted (can't be exposed) and which are committed (can be exposed).

> Control records are not exposed to applications by Kafka consumers, as they are an internal implementation detail of the transaction support.

Reviewer 4: Got it, thanks!

kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
# attributes: Count, 99thPercentile
```

Likewise the following producer metrics can be helpful, perhaps after broker logs reveal some misbehavior.

```sh
kafka.producer:type=producer-metrics,client-id=([-.\w]+)
# attributes: txn-init-time-ns-total, txn-begin-time-ns-total, txn-send-offsets-time-ns-total, txn-commit-time-ns-total, txn-abort-time-ns-total
```

### Basic example

Let's run a basic example to see how transactions work at the partition level.
You just need an environment with few command line tools (curl, tar, java, mvn).

This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.

> The following diagram shows the relevant RPCs.

Author: Ok. I added a row about this.

$ curl -sLk "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.13-$KAFKA_VERSION.tgz" \
  | tar xz -C "$KAFKA_HOME" --strip-components 1

$ $KAFKA_HOME/bin/zookeeper-server-start.sh -daemon $KAFKA_HOME/config/zookeeper.properties \
  && sleep 5 && $KAFKA_HOME/bin/kafka-server-start.sh -daemon $KAFKA_HOME/config/server.properties

# open a new terminal, get and run the application (Ctrl+C to stop)
$ git clone git@github.com:fvaleri/examples.git -n --depth=1 --filter=tree:0 && cd examples \
  && git sparse-checkout set --no-cone kafka/kafka-txn && git checkout 957981f5d8ede82a265f695d9a10c83a50af718f
...
HEAD is now at 957981f Only reset affected partitions

$ export BOOTSTRAP_SERVERS="localhost:9092" INSTANCE_ID="kafka-txn-0" \
  GROUP_ID="my-group" INPUT_TOPIC="input-topic" OUTPUT_TOPIC="output-topic"

$ mvn clean compile exec:java -q -f kafka/kafka-txn/pom.xml
Starting instance kafka-txn-0
Created topics: input-topic
Waiting for new data
...

# come back to the previous terminal and send a test message
$ $KAFKA_HOME/bin/kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
>test
>^C

> ```

Author: I think this is already stated at the start.

> This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.

> ```

Reviewer 4: That's true but you are not showing it working. You just sent a message via the kafka-console-producer not showing the final outcome.

> ```

Author: The final outcome can be seen from logs below, but that's not the point of the example.

```sh
 6 | offset: 0 CreateTime: 1680383687941 keySize: -1 valueSize: 14 sequence: 0 headerKeys: [] payload: tset
```

> ```

Reviewer 4: I know it's not but would help user expectations on trying the example. Anyway not a big deal. Let's keep as it is.

All of this is pretty normal, but now comes the interesting part.

> How can we identify which partitions are involved in this transaction and inspect their content?

Reviewer 4: partitions of the consumer offsets and transaction state topics, right? Isn't it better to clarify? The goal is not pretty clear to me ...

Author: No, the question is about every involved partition. In addition to the ones you mention, there is also the output topic partitions.

> We are writing to the `output-topic` which has only one partition, but the two internal topics have 50 partitions each by default.

Author: It was like this before, but it's repeated just a couple of lines below.

12 Dumping /tmp/kafka-logs/__consumer_offsets-12/00000000000000000000.log
13 Starting offset: 0
14 ...
15 baseOffset: 1 lastOffset: 1 count: 1 baseSequence: 0 lastSequence: 0 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: false deleteHorizonMs: OptionalLong.empty position: 339 CreateTime: 1665506597950 size: 118 magic: 2 compresscodec: none crc: 4199759988 isvalid: true
16 | offset: 1 CreateTime: 1680383688085 keySize: 26 valueSize: 24 sequence: 0 headerKeys: [] key: offset_commit::group=my-group,partition=input-topic-0 payload: offset=1
17 baseOffset: 2 lastOffset: 2 count: 1 baseSequence: -1 lastSequence: -1 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: true deleteHorizonMs: OptionalLong.empty position: 457 CreateTime: 1665506597998 size: 78 magic: 2 compresscodec: none crc: 3355926470 isvalid: true
18 | offset: 2 CreateTime: 1680383688163 keySize: 4 valueSize: 6 sequence: -1 headerKeys: [] endTxnMarker: COMMIT coordinatorEpoch: 0
19 ...
```

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Author: It's the 2PC like process to guarantee that. These end txn markers are appended between `PrepareCommit` and `CompleteCommit` state changes. Once the Producer.commitTransaction is called and `PrepareCommit` is appended to the __transaction_state log, then the transaction coordinator is doomed to commit, not matter what.

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Reviewer 4: Tbh I cannot match your explanation above with the actual output I see from line 8 to 18. These state changes are not visible there, or?

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Author: Ok, maybe as an atomic unit is confusing here, let me remove it.

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Reviewer 4: The thing is not the wording, it's that I don't see what you explain pretty well above ... but in the output lines of the dump. I mean where in the dump I can see this ...

>  These end txn markers are appended between PrepareCommit and CompleteCommit state changes. Once the Producer.commitTransaction is called and PrepareCommit is appended to the __transaction_state log

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Author: You need to look at all the dumps, including the one from `__transaction_state` partition.

> The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

Reviewer 4: well I find doing the correlation not so simple. Not sure it would deserve a few lines to explain it better.

```sh
# requests: FindCoordinator, InitProducerId, AddPartitionsToTxn, Produce, EndTxn
kafka.network:type=RequestMetrics,name=RequestsPerSec,request=([\w]+)
kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
# attributes: Count, 99thPercentile
```

Likewise the following producer metrics can be helpful, perhaps after broker logs reveal some misbehavior.

```sh
kafka.producer:type=producer-metrics,client-id=([-.\w]+)
# attributes: txn-init-time-ns-total, txn-begin-time-ns-total, txn-send-offsets-time-ns-total, txn-commit-time-ns-total, txn-abort-time-ns-total
```

### Basic example

Let's run a basic example to see how transactions work at the partition level.
You just need an environment with few command line tools (curl, tar, java, mvn).

> This application consumes text messages from the `input-topic`, reverts their content, and sends the result to the `output-topic`.

Reviewer 5: reverts -> reverses

Author: Thanks.

12 Dumping /tmp/kafka-logs/__consumer_offsets-12/00000000000000000000.log
13 Starting offset: 0
14 ...
15 baseOffset: 1 lastOffset: 1 count: 1 baseSequence: 0 lastSequence: 0 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: false deleteHorizonMs: OptionalLong.empty position: 339 CreateTime: 1665506597950 size: 118 magic: 2 compresscodec: none crc: 4199759988 isvalid: true
16 | offset: 1 CreateTime: 1680383688085 keySize: 26 valueSize: 24 sequence: 0 headerKeys: [] key: offset_commit::group=my-group,partition=input-topic-0 payload: offset=1
17 baseOffset: 2 lastOffset: 2 count: 1 baseSequence: -1 lastSequence: -1 producerId: 0 producerEpoch: 0 partitionLeaderEpoch: 0 isTransactional: true isControl: true deleteHorizonMs: OptionalLong.empty position: 457 CreateTime: 1665506597998 size: 78 magic: 2 compresscodec: none crc: 3355926470 isvalid: true
18 | offset: 2 CreateTime: 1665506597998 keySize: 4 valueSize: 6 sequence: -1 headerKeys: [] endTxnMarker: COMMIT coordinatorEpoch: 0
19 ...
```

The consumer offsets and the output message are committed as an atomic unit (lines 8 and 18).

> In `wc-output-0` dump, we see that the data batch is transactional and contains the producer's state (line 5).

Reviewer 4: ```suggestion
In `output-topic-0` dump, we see that the data batch is transactional and contains the producer's state (line 5).
```

Reviewer 4: also where is the producer's state?

Author: I meant producer session, as defined above. Fixed.

This is the typical scenario with many stream processing applications that run read-process-write cycles.
In this case, you can make these cycles atomic using the transaction support of the Kafka Producer or Kafka Streams APIs.

The transaction support exposed by the Producer API is low level, and needs to be used with carefully in order to actually get transactional semantics.
For those using Kafka Streams it is much simpler: setting `processing.guarantee=exactly_once_v2` will enable EOS on your existing topology.

### Considerations

Each producer must configure its own static and unique [`transactional.id`](https://kafka.apache.org/documentation/#producerconfigs_transactional.id) to enable the zombie fencing logic, which avoids message duplicates.

The `transactional.id` is used to uniquely identify the same logical producer across process restarts.
In contrast to the idempotent producer, a transactional producer instance will be allocated the same PID (but incremented producer epoch), as any previous instance with the same `transactional.id`. 
Together the PID and producer epoch identify a logical producer session.

> When starting, the producer call `initTransactions` to initialize the producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.

Reviewer: ```suggestion
When starting, the client application calls `Producer.initTransactions` to initialize the producer session, allowing the broker to tidy up state associated with any previous incarnations of this producer, as identified by the `transactional.id`.
```

Thereafter, any existing producers using the same `transactional.id` are considered zombies, and are fenced off as soon as they try to send data.

> Before Kafka 2.6 the `transactional.id` had to be a static encoding of the input partitions in order to avoid ownership transfer between application instances during rebalances, that would invalidate the fencing logic.
> This limitation was hugely inefficient, because an application instance couldn't reuse a single thread-safe producer instance, but had to create one for each input partition. 
> It was fixed with by forcing the producer to send the consumer group metadata along with the offsets to commit ([KIP-447](https://cwiki.apache.org/confluence/display/KAFKA/KIP-447%3A+Producer+scalability+for+exactly+once+semantics)).

To guarantee message ordering, a given producer can have at most one open transaction (they are executed serially).

> Consumers with `isolation.level=read_committed` are not allowed to advance to offsets which include open transactions, while messages from aborted transactions are filtered out.

Reviewer: I think we can improve this by making clear where the responsibilities lie ("not allowed to" is ambgiuous and could be misunderstood as a client programmer responsibility).

```suggestion
Brokers will prevent consumers with `isolation.level=read_committed` from advancing to offsets which include open transactions. Messages from aborted transactions are filtered out within the consumer, and are never observed by the client application.
```

The transaction coordinator is a module running inside a Kafka broker.

> Each coordinator owns a subset of the partitions in the internal `__transaction_state` topic (i.e. the partitions for which its broker is the leader).

Reviewer: I don't really understand the point of this paragraph. Why are we telling the reader about the transaction coordinator _here_? It might make more sense to introduce it when you talk about the PID for transactional producer.

Author: Moved, thanks.

It is important to note that transactions are only supported inside a single Kafka cluster.
If the transaction scope includes processors with externally observable side-effects, you would need to use additional components.
The [transactional outbox pattern](https://microservices.io/patterns/data/transactional-outbox.html) could help in this case.

#### Transaction metadata

In order to implement transactions Kafka brokers need to include extra book-keeping information in logs.
Information about producers and their transactions is stored in the `__transaction_state` log which is appended to by the transaction coordinator.

> Within user logs in addition to the usual batches of user records, partition leaders append control records to track which batches have actually been committed and which rolled back.

Reviewer: ```suggestion
Within user logs in addition to the usual batches of user records, transaction coordinators cause partition leaders append control records to track which batches have actually been committed and which rolled back.
```

- A marker that must be written to each partition when the transaction is completed.
- A few state change records that are appended to the internal transaction state log.

Most of this latency is on the producer side, where the transactions API is implemented.
The consumer is only be able to fetch up to the LSO, but it has to buffer records in memory until the commit marker is observed, so you have increased memory usage.

The performance degradation can be significant when having short transaction commit intervals.
Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.

#### Hanging transactions

Hanging transactions have a missing or out-of-order control record and may be caused by delayed messages due to a network issue ([KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).
This issue is rare, but it's important to be aware and set alerts, because the consequences can impact cluster availability.

The transaction coordinator automatically aborts any ongoing transaction that is not completed within `transaction.timeout.ms`.

> This mechanism does not work for hanging transactions because they are marked as completed.

Reviewer: I wonder if a little more detail helps, wdyt?

```suggestion
This mechanism does not work for hanging transactions because from the coordinator's point of view the transaction was completed and a transaction marker written (and no longer needs to be timed out). But from the partition leader's point of view there is a data record for a transaction after the commit/abort marker, which is indistinguishable from the start of a new transaction.
```
